### PR TITLE
[Feature] Add Cleanlab's Trustworthiness Score 

### PR DESCRIPTION
**Pull request by AshishSardana**
[Cleanlab](https://cleanlab.ai/tlm/) provides trustworthiness score that can be used for guardrailing tasks. Read more about how this score is calculated [here](https://help.cleanlab.ai/tutorials/tlm/#how-does-the-tlm-trustworthiness-score-work).

This PR aims to add support for a 3rd party API i.e. Cleanlab Studio's API, to enable users to use trustworthiness score for using guardrails on output.

### CODE REVIEW

**Comment by AshishSardana** [comment]
Hi @drazvan, I'd like to add brief documentation about trustworthiness score and configurable parameter in the rail (threshold on `least_trustworthiness_score`).
Would it go [here](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/docs/user_guides/guardrails-library.md#third-party-apis)?

Also, how do I add tests? Do you store the keys on the CI server to be able to run the tests?

**Comment by drazvan** [comment]
@AshishSardana : Yes, you can add it in the library documentation. We are in the process of refactoring the structure a bit, but we'll change this as well once that's finalized.

File: nemoguardrails/library/cleanlab/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Cleanlab
+
+This package implements the Cleanlab's [trustworthiness score](https://cleanlab.ai/tlm/) API. <br />
+Feel free to refer to the [announcement blog post](https://cleanlab.ai/blog/trustworthy-language-model/) for more details.
```
Comment (nelsonauner) [inline nemoguardrails/library/cleanlab/README.md]:
```suggestion
Trustworthiness scores help reduce hallucinations and enable reliable deployment of LLMs - learn more in [Overcoming Hallucinations with the Trustworthy Language Model](https://cleanlab.ai/blog/trustworthy-language-model/).
```

File: nemoguardrails/library/cleanlab/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+## Cleanlab Studio
```
Comment (nelsonauner) [inline nemoguardrails/library/cleanlab/README.md]: Our goal in these readme's is to make it as simple as possible for users to get started, so I think something like this is better:

## Development setup

Install `cleanlab-studio` to use Cleanlab's trustworthiness score:

```
pip install cleanlab-studio
```

Then, you can get an API key for free by [creating a Cleanlab account](https://app.cleanlab.ai/?signup_origin=TLM), or experiment with TLM in the [playground](https://tlm.cleanlab.ai/).

File: nemoguardrails/library/cleanlab/actions.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from typing import Optional
+
+from nemoguardrails.actions import action
+from nemoguardrails.colang.v1_0.lang.utils import new_uuid
+
+log = logging.getLogger(__name__)
+
+
+@action(name="call cleanlab api", is_system_action=True)
+async def call_cleanlab_api(context: Optional[dict] = None):
```
Comment (nelsonauner) [inline nemoguardrails/library/cleanlab/actions.py]: Nit: can be helpful to put in return type since you also have input types

**Review by nelsonauner — COMMENTED** [review]
Please take this review with a grain of salt. My comments are mainly to make our documentation more clear and succinct. I also try to facilitate usage of Cleanlab by using phrases like "It's easy to get started by doing X, Y and Z" instead of "It is required to do X if you want to use cleanlab" 

I have some comments that are really more a preference, which I've prefaced with "Nit"

File: nemoguardrails/library/cleanlab/README.md (hunk 2 above)
Comment (AshishSardana) [reply nemoguardrails/library/cleanlab/README.md]: Thanks Nelson! I've improved the text.

File: nemoguardrails/library/cleanlab/actions.py (hunk 3 above)
Comment (AshishSardana) [reply nemoguardrails/library/cleanlab/actions.py]: Added return type.

**Comment by AshishSardana** [comment]
I'm not sure how the commit history got duplicated with my last https://github.com/NVIDIA/NeMo-Guardrails/pull/572/commits/a7517e30bc775f7573ab03ba680d73c34647c8cb commit. 

No new changes are introduced in these duplicate commits.

File: nemoguardrails/library/cleanlab/flows.co [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+https://help.cleanlab.ai/tutorials/tlm/#how-does-the-tlm-trustworthiness-score-work
+
+"""
+
+define subflow cleanlab trustworthiness
+  """Guardrail based on the trustworthiness score."""
+  $result = execute call cleanlab api
+
+  if $result.trustworthiness_score < 0.9
```
Comment (jwmueller) [inline nemoguardrails/library/cleanlab/flows.co]:
```suggestion
  if $result.trustworthiness_score < 0.6
```

Comment (jwmueller) [reply nemoguardrails/library/cleanlab/flows.co]: not sure what impact this has, but I recommend default of 0.6 for the alert threshold

**Comment by AshishSardana** [comment]
Hi @drazvan, the PR is ready for review. Could you please give it a look?

File: docs/user_guides/guardrails-library.md [hunk 5]
```diff
@@ -788,6 +789,38 @@ define flow
...
+NeMo Guardrails supports using the [Cleanlab Trustworthiness Score API](https://cleanlab.ai/blog/trustworthy-language-model/) as an output rail (you need to have the `CLEANLAB_API_KEY` environment variable set).
+
+```yaml
+rails:
+  output:
+    flows:
+      - cleanlab trustworthiness
+```
+
+The `cleanlab trustworthiness` flow uses trustworthiness score with a default threshold of 0.5 to determine if the output should be allowed or not (i.e., if the trustworthiness score is below the threshold, the response is considered "untrustworthy").
```
Comment (Pouyanpi) [inline docs/user_guides/guardrails-library.md]: Should this value be 0.6 instead?

Comment (AshishSardana) [reply docs/user_guides/guardrails-library.md]: thanks for catching that! i've updated the doc.

**Comment by AshishSardana** [comment]
Hi @Pouyanpi , I've resolved the conflicts.

**Comment by AshishSardana** [comment]
Hi @Pouyanpi, please feel free to ask any clarifying questions!

**Comment by drazvan** [comment]
@AshishSardana, I'm just reviewing this, apologies for the delay. The code looks good, however, when I try to run it, I'm running into an error coming from the CleanLab python client. 
Steps to reproduce:
1. Edit the `config.yml` file for the `examples/bots/abc` bot:
```yaml
rails:
#  input:
#    flows:
#      - self check input

  output:
    flows:
      - cleanlab trustworthiness

```
(comment the self-check input and replace the self-check output with cleanlab trustworthiness). 

2. Start the chat CLI `nemoguardrails chat --config=examples/bots/abc --verbose`. 

3. Say "hi"

Relevant error:
```
Traceback (most recent call last):
  File ".../nemoguardrails/nemoguardrails/actions/action_dispatcher.py", line 197, in execute_action
    result = await result
  File ".../nemoguardrails/nemoguardrails/library/cleanlab/actions.py", line 48, in call_cleanlab_api
    trustworthiness_score = await cleanlab_tlm.get_trustworthiness_score_async(
  File ".../nemoguardrails/venv/lib/python3.10/site-packages/cleanlab_studio/studio/trustworthy_language_model.py", line 559, in get_trustworthiness_score_async
    assert isinstance(prompt, Sequence) and isinstance(processed_response, Sequence)
AssertionError
```

From what I can see, the `processed_response` is actually a dict like `{"response": "..."}`. 

Let me know if you're able to reproduce

**Comment by AshishSardana** [comment]
@drazvan I was able to reproduce this error. We've pushed a fix to the cleanlab-studio package that you can fetch by `pip install -U cleanlab-studio` which resolves this error. Could you try?

**Review by drazvan — APPROVED** [review]
I was able to test and looks good :+1:. Latency was a bit high in my tests, ~5s, but I guess that will be improved.

**Comment by cgnorthcutt** [comment]
@drazvan and team, latency of https://cleanlb.ai/tlm has been improved (typically within 1-2 seconds of when output ends for moderate length outputs)

**Pull request merged**